Add authentication support to Airflow connectors

ptitzler opened the pull request:
This PR
 - adds optional authentication support to the Airflow package catalog connector
 - adds optional authentication support to the Airflow provider package catalog connector
 - updates the relevant documentation.
 
Existing connector configurations will continue to work as is.

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our [contributor guidelines](https://github.com/elyra-ai/community/blob/main/contributing.md)
    1.1 Please follow the [7 rules for a great commit message](https://github.com/elyra-ai/community/blob/main/contributing.md#creating-a-pull-request)
  2. If the PR is unfinished, leave it as 'DRAFT', add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...' or use the 'status:Work in Progress' label.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If this PR involves UI changes, please provide a screen capture (before/after) for a faster review.
  6. Remember to add 'Fixes #ISSUE_NUMBER' (or any [valid keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)) to the end of your message body to get the issue properly closed when the PR is merged.
  7. Set the proper milestone based on the target release for the issue.
  8. Consider whether any documentation need to be added or updated based on your changes.
-->

### What changes were proposed in this pull request?

- Extend Airflow package connector configuration to accept optional credentials
- Extend Airflow provider package connector configuration to accept optional credentials

![image](https://user-images.githubusercontent.com/13068832/175313050-c7a8abb7-5ec3-4dd8-aa4f-89bbd56b920d.png)

- A valid set of credentials must be provided. If only a user id or only password is entered, the connectors will log an error in the console.

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor code that leads to class changes, showing the updated class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
-->

### How was this pull request tested?
 - Configure the Airflow package catalog connector to download an Airflow package from an unsecured web location:
    - configure the connector to not include any credentials (should succeed)
    - configure the connector to include bogus credentials (should succeed)
 - Configure the Airflow package catalog connector to download an Airflow package from a secured web location
    - provide valid user id and password (should succeed)
    - provide invalid user id/password combination (should fail)
    - provide only a user id or password (should fail)
 - Configure the Airflow provider package catalog connector to download an Airflow package from an unsecured web location:
    - configure the connector to not include any credentials (should succeed)
    - configure the connector to include bogus credentials (should succeed)
 - Configure the Airflow provider package catalog connector to download an Airflow package from a secured web location
    - provide valid user id and password (should succeed)
    - provide invalid user id/password combination (should fail)
    - provide only a user id or password (should fail)
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly
including negative and positive cases if possible.

If it was tested in a way different from regular unit tests, please clarify how you tested step by step,
ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.

If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ptitzler commented:
@shalberd I've tested this with a couple artifacts that I've deployed in a local artifactory deployment. Does the proposed solution address your needs?

shalberd commented on elyra/pipeline/airflow/package_catalog_connector/README.md:
```diff
@@ -10,7 +10,12 @@ https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#manag
 1. Specify a catalog name, e.g. '`Airflow 1.10.15 wheel`'.
 1. (Optional) Specify a category under which the loaded operators will be organized in the palette.
-1. Configure the '`Airflow package download URL`'. The URL must reference a location that Elyra can access using an HTTP GET request, without the need to authenticate. If the Airflow package is stored on PyPI:
+1. Configure the '`Airflow package download URL`'. The URL must reference a location that Elyra can access using an HTTP GET request. If the resource is secured, provide credentials, such as a user id or api key and password.
```
such as user id and API key or password

shalberd commented on elyra/pipeline/airflow/provider_package_catalog_connector/airflow-provider-package-catalog.json:
```diff
@@ -61,6 +61,25 @@
+        },
+        "auth_password": {
+          "title": "Password",
+          "description": "Password for specified id",
```
Password or API key for the specified id

shalberd commented on elyra/pipeline/airflow/package_catalog_connector/airflow-package-catalog.json:
```diff
@@ -56,27 +56,29 @@
         "auth_password": {
           "title": "Password",
-          "description": "Password for download user",
+          "description": "Password for specified id",
```
Password or API key for specified id.

Overall, both in the popup hints and code and GUI, we might also think of calling it either user id or username.

shalberd commented on elyra/pipeline/airflow/provider_package_catalog_connector/README.md:
```diff
@@ -10,7 +10,11 @@ https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#manag
 1. Specify a catalog name, e.g. '`HTTP provider package`'.
 1. (Optional) Specify a category under which the loaded operators will be organized in the palette.
-1. Configure the '`Provider package download URL`'. The URL must reference a location that Elyra can access using an HTTP GET request, without the need to authenticate. If the Airflow provider package is stored on PyPI:
+1. Configure the '`Provider package download URL`'. If the resource is secured, provide credentials, such as a user id or api key and password.
```
such as a user id and API key or password.

shalberd commented on elyra/pipeline/airflow/package_catalog_connector/airflow_package_catalog_connector.py:
```diff
@@ -67,21 +67,23 @@ def get_catalog_entries(self, catalog_metadata: Dict[str, Any]) -> List[Dict[str
-                "Authentication requires a user and password."
+                f"Error. Airflow connector '{catalog_metadata.get('display_name')}' "
+                "is not configured properly. "
+                "Authentication requires an id and password."
```
a (user) id and password or API key

shalberd commented:
Looking good overall to me. Main naming in my view should be "user id" instead of id in both GUI and code. And the authentication is always a combination of user id and password or API key. API key is basically a aubstitute for password, not for username.

kevin-bates commented on elyra/pipeline/airflow/package_catalog_connector/airflow-package-catalog.json:
```diff
@@ -56,11 +56,30 @@
+          "description": "User id that has read access for the specified URL resource",
+          "type": "string",
+          "uihints": {
+            "minLength": 1,
```
I think `minLength` should be a direct attribute of `"auth_id"` (and `"auth_password"`) rather than a ui-hint - assuming you want validation to catch empty strings when defined.

kevin-bates commented on elyra/pipeline/airflow/package_catalog_connector/airflow-package-catalog.json:
```diff
@@ -56,11 +56,30 @@
           "title": "Airflow package download URL",
-          "description": "URL where the Apache Airflow package wheel can be downloaded.",
+          "description": "URL where the Apache Airflow package wheel can be downloaded",
           "type": "string",
           "uihints": {
```
I think we should add a URI format specifier here so validation can ensure a valid format...
```suggestion
          "type": "string",
          "format": "uri",
          "uihints": {
```

kevin-bates commented on elyra/pipeline/airflow/provider_package_catalog_connector/airflow-provider-package-catalog.json:
```diff
@@ -61,6 +61,25 @@
+          "description": "User id that has read access for the specified URL resource",
+          "type": "string",
+          "uihints": {
+            "minLength": 1,
```
Same comments as on the previous schema file relative to `minLength` location and the addition of a URI format specifier on the url property.

kevin-bates reviewed (commented):
The changes look good.  Just had some comments relative to the schema definitions.

ptitzler commented:
Addressed review feedback and added placeholder values to URI properties, e.g 

![image](https://user-images.githubusercontent.com/13068832/175570771-e1a53496-9102-4115-afdd-19671e79edbf.png)

This should lower the likelihood that the `uri` format error is raised:

![image](https://user-images.githubusercontent.com/13068832/175570540-efea21d9-2bbe-4b51-8323-51daabbcd45c.png)

kiersten-stokes reviewed (commented):
So I'm having a weird issue where the changes to the Airflow package connector schema aren't getting pulled into the UI (provider package is working fine, as are any changes I make to other schemas). This is in my python 3.9 env.

When I test this in a python 3.7 env, however, I see all the changes. I'm not sure if this is just me, but regardless, I haven't gotten to the bottom of it yet 😄 

### Python 3.7:
<img width="1184" alt="Screen Shot 2022-06-24 at 1 39 00 PM" src="https://user-images.githubusercontent.com/31816267/175645515-6f29a966-f73f-4e9b-b0ac-c77320580f62.png">


### Python 3.9:
<img width="1131" alt="Screen Shot 2022-06-24 at 1 38 50 PM" src="https://user-images.githubusercontent.com/31816267/175645541-7e753398-3f0d-42ed-ab5e-f4f4cc8bd5b3.png">

kiersten-stokes reviewed (approved):
Phew, looks like it was luckily just my env! Everything LGTM

kevin-bates reviewed (approved):
Looks good - thanks @ptitzler!

shalberd commented on elyra/pipeline/airflow/package_catalog_connector/airflow-package-catalog.json:
```diff
@@ -60,7 +60,8 @@
           "uihints": {
-            "category": "Source"
+            "category": "Source",
+            "ui:placeholder": "http://host:port/path/apache_airflow.whl"
```
did not get to this until now, sorry. In the future, maybe you can change the placeholder to https, as is mentioned in the documentation with regards to pythonhosted and as most e.g. Artifactory users will use. Other than that, very good idea with the UI placeholders.

shalberd commented:
UI placeholders for Url should have https in them, as that is standard. Very nice work. With regards to Https, what is your opinion on https://github.com/elyra-ai/elyra/issues/2797 ? Have a great week.

ptitzler commented:
I am working on a separate PR for https://github.com/elyra-ai/elyra/issues/2793 that closes this gap for the URL connector.
